Address late code reviews

aldelaro5 opened the pull request:
@slxdy sent several code reviews and this PR partially addresses them.

Here's all the changes:

#871 Addressed the warning via pragma suppression (proper fix would be enabling nullable, but this is simply not feasible without a complete rewrite of several apis and the deletion of obsoleted ones)
#875 Moved the Mono specific entrypoint to its own class that Core uses
#877 Did nothing for this one because it's not possible to address: the MSBuild changes are required even if one specify the exports file to the linker because .net still mangles them with versions we don't want on Linux
#855 Fixed the bad "--" on the capture player logs CLI arg. As for the PltHook stuff, I didn't do anything because I don't understand the review comment: PltHook is NOT an alternative to dobby because it is more specialised and can't be used for general purpose hooks which is what dobby is for. The way we use PltHook is meant ONLY for the bootstrap so I don't understand why it needs to have a matching api...it doesn't really make sense to me
#887 For the comments on the runtime handlers, I decided to split the dictionnary into 2 where each handlers have their parts and the common RedirectSymbol is the one that consults the 2 from the runtimes. That's essentially the best I can do to leave runtime specific stuff out of core without code duplications. As for the HookGetProcAddress, nothing was done because trying this reveals that NativeLibrary.GetExport is semantically incompatible with dlsym (it doesn't understand that a null handle means "the main library")
#890 Removed the "--" on the harmony log level CLI arg
#900 Simplified the base dir obtention on MacOS...but not in the suggested ways because it doesn't work (doing a combine on .. doesn't do anything). Instead, I made a method that simplifies reading the code, but it does the same semantically
#905 Reverted the commit, but added a check for wine instead as suggested. I tested both on Windows and Linux with Wine/Proton and this check seemed to work. I also verified indeed that reverting was necessary for powershell redirection to work

Note for @HerpDerpinstine : As mentioned in #871 , the ColorRGB breaking change needs to be mentioned somewhere in the next version's release notes.

That should cover most things

aldelaro5 committed:
Suppress a nullability warning in ColorRGB

aldelaro5 committed:
Move the Mono specific entrypoint to its own class

aldelaro5 committed:
Fix some CLI parsing arguments

aldelaro5 committed:
Split the symbols redirects to their runtime handlers

aldelaro5 committed:
Revert "Call AllocConsole even if stdout handle is not null"

This reverts commit 6dc024fda3c8de7203f18f32ccb1850a76c87b74.

aldelaro5 committed:
Don't prevent console window on Windows if running under wine

aldelaro5 committed:
Simplify getting the base dir on MacOS

slxdy commented:
For #900, I think I was high when writing the code lol. My solution was valid, but the `baseDir` arg should've been the first, followed by the multiple `..`

slxdy commented:
For #887, I don't understand why you even need the function left in Core in the first place. There is no point in even storing the target functions in a Dictionary. You can just apply PLT hooks one for one in each runtime handler.

slxdy commented:
For #855, the PLT hooks should be revertible for user convenience (and in this case the user would be another bootstrap dev). For this reason, our Dobby wrapper API is the way it is, and that's why I wanna see something similar for PLT. Obviously it doesn't have to be the same, but it should have a similar structure.
Another thing I mentioned in the PltHook wrapper is that it shouldn't be specific to a single library.

The main reason for all this is that each class should have its own clear purpose. The purpose for the Dobby class is to be a simple .net wrapper for the Dobby library, which utilizes the features of .net to make the API more convenient than just exposing the imported functions.
PltHook should follow the same pattern. Right now, the PltHook wrapper breaks its purpose by making it specific to a single library and only exposing a function meant for specific cases.

While yes, technically all public API in the bootstrap is not really "public", it should still be considered as that. This way future developers don't have to worry about "What does this class even hook?" or "How do I revert hooks through this class?"

slxdy commented:
The reason why I'm being so strict about it is because I don't want the bootstrap API to end up like the managed MelonLoader API, where half of the API makes no sense and you have to actually read the internal code to understand what anything does.

aldelaro5 commented:
> For #887, I don't understand why you even need the function left in Core in the first place. There is no point in even storing the target functions in a Dictionary. You can just apply PLT hooks one for one in each runtime handler.

No because you cannot know in advance if the function exists.

That's the ENTIRE point of soft hooking like this: you don't need to deal with "does the symbol / runtime applies?" and just give a list of known symbols you're interested in to a dlsym hook. There's no probing on anything because unity itself now tells you the hooks you need.

The function needs to be there to avoid code duplication because as a result of this, both runtimes soft hooks in the same way, but they differ in their symbols. You wanted the runtime specific parts to be separated...well I did.

aldelaro5 commented:
> The reason why I'm being so strict about it is because I don't want the bootstrap API to end up like the managed MelonLoader API, where half of the API makes no sense and you have to actually read the internal code to understand what anything does.

Okay, but melon is also a public facing API while the bootstrap has very few things it exposes to melon.

The reason I had arranged the plthook stuff as it is is because before I did this, I noticed I kept using the same patterns of wanting to plthook a list of stuff over and over. I extracted the logic to a class so I avoid code duplication. It wasn't born out of needing a proper wrapper because I didn't see the need for one. I also hadn't tested if the plt hook can be reverted.

What I am saying is if you REALLY want that I split up this class, it will in my opinion unnecessarily complicate things.

aldelaro5 commented:
> For #900, I think I was high when writing the code lol. My solution was valid, but the `baseDir` arg should've been the first, followed by the multiple `..`

I'll retest once I sort out the net10 investigation (macos testing takes a while).

slxdy commented on MelonLoader.Bootstrap/Logging/ConsoleHandler.cs:
```diff
@@ -75,6 +77,16 @@ public static void OpenConsole(bool onTop, string? title)
+        return isWine;
+    }
+#endif
+
```
This check is already present in WineUtils btw

aldelaro5 replied:
Done.

aldelaro5 committed:
Use existing method to check for wine in the ConsoleHandler
